### PR TITLE
Makefile and company

### PR DESCRIPTION
**Pull request by Pixailz**
There's a new [Makefile](https://github.com/Pixailz/ft_transcendence/blob/main/Makefile) and a [setup](https://github.com/Pixailz/ft_transcendence/blob/main/setup) file to help us starting up the project

The Makefile help with choosing the prod and dev version of our project
- `make` will launch it for prod while `make dev` will launch it for us
  - `docker-compose.yaml` -> prod
  - `docker-compose.dev.yaml` -> dev (for us)

The setup file is just here to help us filling the .env if not there

no other file impacted

**IMPORTANT**: we have to define what an dev env and prod env

### CODE REVIEW

**Comment by clinche** [comment]
> we have to define what an dev env and prod env

tl;dr picture
this is only true for angular, which give us the possibility to build for dev or for prod. it's quite like the equivalent of using -g3 on binaries vs stripping them. from what i get, nestjs already is a prod server (to confirm)

in a dev environment, we would have the angular dev server handle every request and proxy api to the nestjs backend

in a prod environment, we would then have all angular compiled ts files, a nestjs backend listening on a docker local network on port 3000, and an nginx reverse proxy exposed on port 443. this reverse proxy would be responsible for serving the actual compiled angularjs website, and would also be responsible for forwarding every request starting with `/api` to the nestjs backend
image

**Review by clinche — COMMENTED** [review]
clearly there's more to do if we wanna get a dev and a prod env, but this is great beginning

**Comment by clinche** [comment]
The front-end won't be enough because in prod there'll be no front end server
The angular dev server compiles and serve typescript files on the fly, but if we put angular in prod mode it will only compile typescript and produce a servable "static" website, that we need to serve with a real web server

But nestjs **apparently** already have a prod server, thus we need nginx in a reverse proxy mode to balance between angular already compiled files and nestjs running server

**Comment by Pixailz** [comment]
I implemented a way to switch between prod and dev, in the [.env](https://github.com/Pixailz/ft_transcendence/blob/24-bash/.env.template) file, if `DEBUG` is null or unset, it will launch the prod rule in the package.json (in both, [angular](https://github.com/Pixailz/ft_transcendence/blob/24-bash/src/angularjs/package.json) and [nestjs](https://github.com/Pixailz/ft_transcendence/blob/24-bash/src/nestjs/package.json))

but i don't know how to configure it to make it reverse proxying between angular or nestjs

what do you think of this features ?

**Review by clinche — CHANGES_REQUESTED** [review]
thats sound great, this should works fine
i commented the needed code for nginx, this should do it, we'll just have to be careful when we'll need websockets and X-Real-IP header as this could be spoofed by client (?)

File: build/nginx/default.conf [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+	error_log		/var/log/nginx/error.log debug;
+	access_log		/var/log/nginx/access.log;
+
+	root /var/www/html;
+
+	location / {
+		try_files $uri $uri/ =404;
+	}
+	location /api {
+
```
[the remark this answers is not in the document]
Comment (Pixailz) [reply build/nginx/default.conf:21]: i think the client can spoof it's real IP with a VPN but we don't make any security check with the ip so, i think it cannot be a threat

**Comment by Pixailz** [comment]
also, on the angularjs side, do we have to call `ng serve` or only `ng build` ?
like the angularjs service will just build the typescript and nginx will serv the compiled file right ?

File: build/nginx/default.conf (hunk 1 above)
[the remark this answers is not in the document]
Comment (clinche) [reply build/nginx/default.conf:21]: no im talking like header poisoning
[the docs]() says this

>Allows redefining or appending fields to the request header [passed](http://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_pass_request_headers) to the proxied server. The value can contain text, variables, and their combinations. These directives are inherited from the previous configuration level if and only if there are no proxy_set_header directives defined on the current level. By default, only two fields are redefined:
    `proxy_set_header Host       $proxy_host;`
    `proxy_set_header Connection close;`

so maybe if a user sends a X-Real-IP with a random IP nginx will just append `$remote_addr`, so we would need to validate if X-Real-IP contains only one IP

**Comment by clinche** [comment]
> angularjs service will just build the typescript and nginx will serv the compiled file right ?

yep totally, `ng build` but nginx serves :p

File: build/nginx/default.conf (hunk 1 above)
[the remark this answers is not in the document]
Comment (Pixailz) [reply build/nginx/default.conf]: this method don't work, have to decide how this can be done
